sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		return nil, 0, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.InitQuoteExReq = &pb.AesmServiceRequest_InitQuoteExRequest{
+		AttKeyId:         akId,
+		BufSize:          bufSize,
+		BPubKeyIdPresent: &pb.AesmServiceRequest_InitQuoteExRequest_BPubKeyId{BPubKeyId: akPubKeyRequired},
+		Timeout:          aesmdTimeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.InitQuoteExRes = &pb.AesmServiceResponse_InitQuoteExResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return nil, 0, err
+	}
+
+	if errCode := resp.InitQuoteExRes.GetErrorCode(); errCode != 0 {
+		return nil, 0, fmt.Errorf("failed to call initQuoteExRequest (error code = %d)", errCode)
+	}
+
+	return resp.InitQuoteExRes.GetTargetInfo(), resp.InitQuoteExRes.GetPubKeyIdSize(), nil
+}
+
+func getQuoteSizeExRequest(akId []byte) (uint32, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return 0, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.GetQuoteSizeExReq = &pb.AesmServiceRequest_GetQuoteSizeExRequest{
+		AttKeyId: akId,
+		Timeout:  aesmdTimeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return 0, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetQuoteSizeExRes = &pb.AesmServiceResponse_GetQuoteSizeExResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return 0, err
+	}
+
+	if errCode := resp.GetQuoteSizeExRes.GetErrorCode(); errCode != 0 {
+		return 0, fmt.Errorf("failed to call getQuoteSizeExRequest (error code = %d)", errCode)
+	}
+
+	return resp.GetQuoteSizeExRes.GetQuoteSize(), nil
+}
+
+func getQuoteExRequest(akId []byte, report []byte, quoteSize uint32) ([]byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.GetQuoteExReq = &pb.AesmServiceRequest_GetQuoteExRequest{
+		Report:       report,
+		AttKeyId:     akId,
+		QeReportInfo: nil,
+		BufSize:      quoteSize,
+		Timeout:      aesmdTimeOut,
+	}
+
+	rdata, err := transmitAesmd(conn, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := pb.AesmServiceResponse{}
+	resp.GetQuoteExRes = &pb.AesmServiceResponse_GetQuoteExResponse{}
+	if err := proto.Unmarshal(rdata, &resp); err != nil {
+		return nil, err
+	}
+
+	if errCode := resp.GetQuoteExRes.GetErrorCode(); errCode != 0 {
+		return nil, fmt.Errorf("failed to call getQuoteExRequest (error code = %d)", errCode)
+	}
+
+	return resp.GetQuoteExRes.GetQuote(), nil
+}
+
+func selectAttKeyIDRequest() ([]byte, error) {
+	conn, err := dialAesmd()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	req := pb.AesmServiceRequest{}
+	req.SelectAttKeyIDReq = &pb.AesmServiceRequest_SelectAttKeyIDRequest{
+		// AttKeyIdList is optional for aesmd so that we don't need to fill in the corresponding Attestation Key Identity List
+		AttKeyIdList: nil,
```

Comment:
You are explaining nothing. It sounds like "null pointer indicates an invalid address". You need to explain why nil is used.